aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu

```diff
@@ -390,26 +391,51 @@ SparseTensor& add_out_sparse_cuda(const SparseTensor& t, const SparseTensor& src
   auto commonDtype = at::result_type(t, src);
   TORCH_CHECK(canCast(commonDtype, r_.scalar_type()), "Can't convert result type ", commonDtype, " to output ", r_.scalar_type());
 
-  TORCH_CHECK(t.sizes().equals(src.sizes()), "add: expected 'self' and 'other' to have same size, but ", t.sizes(), " != ", src.sizes());
+  // Check if the broadcasting can be performed and if applicable return the shape of the result.
+  const std::vector<int64_t>& res_shape = infer_size(t.sizes(), src.sizes());
+  bool is_same_size = t.sizes().equals(src.sizes());
 
+  // deal with empty sparse tensors
   if (src._nnz() == 0) {
-    return copy_sparse_to_sparse_(r_, t);
+    if (is_same_size) {
+      return copy_sparse_to_sparse_(r_, t);
+    } else {
+      const SparseTensor& broadcasted_t = sparse_broadcast_to(t, res_shape);
+      return copy_sparse_to_sparse_(r_, broadcasted_t);
+    }
   }
   if (t._nnz() == 0) {
-    return mul_out_sparse_scalar(r_, src, value);
+    if (is_same_size) {
+      return mul_out_sparse_scalar(r_, src, value);
+    } else {
+      const SparseTensor& broadcasted_src = sparse_broadcast_to(src, res_shape);
+      return mul_out_sparse_scalar(r_, broadcasted_src, value);
+    }
   }
 
-  TORCH_CHECK(is_same_density(t, src), "add: expected 'self' and 'other' to have same density, but 'self' has ", t.sparse_dim(), " sparse dimensions while 'other' has ", src.sparse_dim(), " sparse dimensions");
+  // the two sparse tensors should have the same dense_dim.
+  bool is_same_dense_ndim = (t.dense_dim() == src.dense_dim());
+  TORCH_CHECK(is_same_dense_ndim, "add: expected 'self' and 'other' to have same number of dense dimensions, but 'self' has ", t.dense_dim(), " dense dimensions while 'other' has ", src.dense_dim(), " dense dimensions");
 
   // We deliberately choose to simply concat the indices and values tensors
   // rather than merging them. This removes the need to synchronously fetch nnz
   // at the end of the operation, at the cost of having a non-coalesced result.
   // This trade-off is preferable for the common use-case of gradient accumulation.
-  Tensor t_indices_ = t._indices();
-  Tensor s_indices_ = src._indices();
+  SparseTensor broadcasted_t;
+  SparseTensor broadcasted_src;
+  if (is_same_size) {
+    broadcasted_t = t;
+    broadcasted_src = src;
+  } else {
+    broadcasted_t = sparse_broadcast_to(t, res_shape);
+    broadcasted_src = sparse_broadcast_to(src, res_shape);
+  }
```

Comment:
The logic repeats itself. Why not just always broadcast with these ops? Then the rest of the code will stay unchanged.